crates/types/src/traits/block_contents.rs

```diff
@@ -50,7 +46,17 @@ pub trait BlockPayload:
     /// The type of the transitions we are applying
     type Transaction: Transaction;
 
-    // type Header: BlockHeader;
+    /// Data created during block building which feeds into the block header
+    type Metadata;
+
+    /// Build a payload and associated metadata with the transactions.
+    fn build(transactions: impl IntoIterator<Item = Self::Transaction>) -> (Self, Self::Metadata);
+
+    /// Encode the payload
+    fn encode(&self) -> Vec<u8>;
+
+    /// Decode the payload
+    fn decode(encoded_transactions: &[u8]) -> Self;
```

Comment:
I think `decode` has to take the `Metadata` as well. E.g. in the sequencer, we can't find the transaction boundaries without the namespace table.